File: internal/orchestrator/tools_admin.go
package orchestrator

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/Z-M-Huang/openhive/internal/domain"
)

// AdminToolsDeps holds dependencies for admin tool handlers.
type AdminToolsDeps struct {
	ConfigLoader domain.ConfigLoader
	KeyManager   domain.KeyManager
	WSHub        domain.WSHub
	StartTime    time.Time
}

// RegisterAdminTools registers all admin SDK custom tool handlers on the ToolHandler.
func RegisterAdminTools(handler *ToolHandler, deps AdminToolsDeps) {
	handler.Register("get_config", makeGetConfig(deps.ConfigLoader))
	handler.Register("update_config", makeUpdateConfig(deps.ConfigLoader))
	handler.Register("get_system_status", makeGetSystemStatus(deps))
	handler.Register("list_channels", makeListChannels(deps.WSHub))
	handler.Register("enable_channel", makeEnableChannel(deps.ConfigLoader))
	handler.Register("disable_channel", makeDisableChannel(deps.ConfigLoader))
}

// getConfigArgs holds arguments for get_config tool.
type getConfigArgs struct {
	Section string `json:"section"`
}

func makeGetConfig(configLoader domain.ConfigLoader) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		var a getConfigArgs
		if err := json.Unmarshal(args, &a); err != nil {
			return nil, &domain.ValidationError{Field: "args", Message: "invalid arguments"}
		}

		cfg, err := configLoader.LoadMaster()
		if err != nil {
			return nil, fmt.Errorf("failed to load config: %w", err)
		}

		var section interface{}
		switch a.Section {
		case "system":
			section = cfg.System
		case "assistant":
			section = cfg.Assistant
		case "channels":
			// Redact sensitive fields
			redacted := cfg.Channels
			if redacted.Discord.Token != "" {
				redacted.Discord.Token = "[REDACTED]"
			}
			if redacted.WhatsApp.Token != "" {
				redacted.WhatsApp.Token = "[REDACTED]"
			}
			section = redacted
		case "":
			// Return entire config with redaction
			cfgCopy := *cfg
			if cfgCopy.Channels.Discord.Token != "" {
				cfgCopy.Channels.Discord.Token = "[REDACTED]"
			}
			if cfgCopy.Channels.WhatsApp.Token != "" {
				cfgCopy.Channels.WhatsApp.Token = "[REDACTED]"
			}
			section = cfgCopy
		default:
			return nil, &domain.ValidationError{Field: "section", Message: fmt.Sprintf("unknown section: %s", a.Section)}
		}

		return json.Marshal(section)
	}
}

// updateConfigArgs holds arguments for update_config tool.
type updateConfigArgs struct {
	Section string      `json:"section"`
	Field   string      `json:"field"`
	Value   interface{} `json:"value"`
}

func makeUpdateConfig(configLoader domain.ConfigLoader) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		var a updateConfigArgs
		if err := json.Unmarshal(args, &a); err != nil {
			return nil, &domain.ValidationError{Field: "args", Message: "invalid arguments"}
		}

		if a.Section == "" || a.Field == "" {
			return nil, &domain.ValidationError{Field: "section/field", Message: "section and field are required"}
		}

		cfg, err := configLoader.LoadMaster()
		if err != nil {
			return nil, fmt.Errorf("failed to load config: %w", err)
		}

		// Apply update based on section and field
		switch a.Section {
		case "system":
			if err := applySystemUpdate(cfg, a.Field, a.Value); err != nil {
				return nil, err
			}
		case "channels":
			if err := applyChannelUpdate(cfg, a.Field, a.Value); err != nil {
				return nil, err
			}
		default:
			return nil, &domain.ValidationError{
				Field:   "section",
				Message: fmt.Sprintf("unsupported section for update: %s", a.Section),
			}
		}

		// Save config (will trigger fsnotify/hot-reload)
		if err := configLoader.SaveMaster(cfg); err != nil {
			return nil, fmt.Errorf("failed to save config: %w", err)
		}

		return json.Marshal(map[string]string{"status": "updated"})
	}
}

func applySystemUpdate(cfg *domain.MasterConfig, field string, value interface{}) error {
	switch field {
	case "log_level":
		v, ok := value.(string)
		if !ok {
			return &domain.ValidationError{Field: "value", Message: "log_level must be a string"}
		}
		cfg.System.LogLevel = v
	case "listen_address":
		v, ok := value.(string)
		if !ok {
			return &domain.ValidationError{Field: "value", Message: "listen_address must be a string"}
		}
		cfg.System.ListenAddress = v
	default:
		return &domain.ValidationError{
			Field:   "field",
			Message: fmt.Sprintf("unsupported system field: %s", field),
		}
	}
	return nil
}

func applyChannelUpdate(cfg *domain.MasterConfig, field string, value interface{}) error {
	switch field {
	case "discord.enabled":
		v, ok := value.(bool)
		if !ok {
			return &domain.ValidationError{Field: "value", Message: "discord.enabled must be a boolean"}
		}
		cfg.Channels.Discord.Enabled = v
	case "whatsapp.enabled":
		v, ok := value.(bool)
		if !ok {
			return &domain.ValidationError{Field: "value", Message: "whatsapp.enabled must be a boolean"}
		}
		cfg.Channels.WhatsApp.Enabled = v
	default:
		return &domain.ValidationError{
			Field:   "field",
			Message: fmt.Sprintf("unsupported channel field: %s", field),
		}
	}
	return nil
}

// systemStatusResult holds the result of get_system_status tool.
type systemStatusResult struct {
	ConnectedTeams []string `json:"connected_teams"`
	Uptime         string   `json:"uptime"`
	Version        string   `json:"version"`
}

func makeGetSystemStatus(deps AdminToolsDeps) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		teams := deps.WSHub.GetConnectedTeams()
		uptime := time.Since(deps.StartTime).Round(time.Second)

		result := systemStatusResult{
			ConnectedTeams: teams,
			Uptime:         uptime.String(),
			Version:        "0.1.0",
		}

		return json.Marshal(result)
	}
}

func makeListChannels(wsHub domain.WSHub) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		teams := wsHub.GetConnectedTeams()
		return json.Marshal(map[string]interface{}{
			"connected_teams": teams,
		})
	}
}

// enableChannelArgs holds arguments for enable_channel tool.
type enableChannelArgs struct {
	Channel string `json:"channel"`
}

func makeEnableChannel(configLoader domain.ConfigLoader) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		var a enableChannelArgs
		if err := json.Unmarshal(args, &a); err != nil {
			return nil, &domain.ValidationError{Field: "args", Message: "invalid arguments"}
		}

		if a.Channel == "" {
			return nil, &domain.ValidationError{Field: "channel", Message: "channel name is required"}
		}

		cfg, err := configLoader.LoadMaster()
		if err != nil {
			return nil, fmt.Errorf("failed to load config: %w", err)
		}

		switch a.Channel {
		case "discord":
			cfg.Channels.Discord.Enabled = true
		case "whatsapp":
			cfg.Channels.WhatsApp.Enabled = true
		default:
			return nil, &domain.ValidationError{
				Field:   "channel",
				Message: fmt.Sprintf("unknown channel: %s", a.Channel),
			}
		}

		if err := configLoader.SaveMaster(cfg); err != nil {
			return nil, fmt.Errorf("failed to save config: %w", err)
		}

		return json.Marshal(map[string]string{"status": "enabled", "channel": a.Channel})
	}
}

func makeDisableChannel(configLoader domain.ConfigLoader) ToolFunc {
	return func(args json.RawMessage) (json.RawMessage, error) {
		var a enableChannelArgs // same args
		if err := json.Unmarshal(args, &a); err != nil {
			return nil, &domain.ValidationError{Field: "args", Message: "invalid arguments"}
		}

		if a.Channel == "" {
			return nil, &domain.ValidationError{Field: "channel", Message: "channel name is required"}
		}

		cfg, err := configLoader.LoadMaster()
		if err != nil {
			return nil, fmt.Errorf("failed to load config: %w", err)
		}

		switch a.Channel {
		case "discord":
			cfg.Channels.Discord.Enabled = false
		case "whatsapp":
			cfg.Channels.WhatsApp.Enabled = false
		default:
			return nil, &domain.ValidationError{
				Field:   "channel",
				Message: fmt.Sprintf("unknown channel: %s", a.Channel),
			}
		}

		if err := configLoader.SaveMaster(cfg); err != nil {
			return nil, fmt.Errorf("failed to save config: %w", err)
		}

		return json.Marshal(map[string]string{"status": "disabled", "channel": a.Channel})
	}
}

File: backend/src/orchestrator/tools-admin.ts
/**
 * OpenHive Backend - Admin SDK Tool Handlers
 *
 * Registers admin tool handlers on the ToolHandler for use by the main
 * assistant. Handlers cover config read/write, system status, and channel
 * enable/disable.
 */

import type { ConfigLoader, KeyManager, WSHub } from '../domain/interfaces.js';
import type { JsonValue, MasterConfig } from '../domain/types.js';
import { AccessDeniedError, ValidationError } from '../domain/errors.js';
import type { ToolFunc, ToolCallContext } from './toolhandler.js';
import type { ToolRegistry } from '../domain/interfaces.js';

// ---------------------------------------------------------------------------
// AdminToolsDeps
// ---------------------------------------------------------------------------

/**
 * Dependencies injected into admin tool handlers.
 */
export interface AdminToolsDeps {
  configLoader: ConfigLoader;
  keyManager: KeyManager;
  wsHub: WSHub;
  startTime: Date;
}

// ---------------------------------------------------------------------------
// registerAdminTools
// ---------------------------------------------------------------------------

/**
 * Registers all admin SDK custom tool handlers on the ToolHandler.
 */
export function registerAdminTools(handler: ToolRegistry, deps: AdminToolsDeps): void {
  handler.register('get_config', withAdminGuard('get_config', makeGetConfig(deps.configLoader)));
  handler.register('update_config', withAdminGuard('update_config', makeUpdateConfig(deps.configLoader)));
  handler.register('get_system_status', withAdminGuard('get_system_status', makeGetSystemStatus(deps)));
  handler.register('list_channels', withAdminGuard('list_channels', makeListChannels(deps.wsHub)));
  handler.register('enable_channel', withAdminGuard('enable_channel', makeEnableChannel(deps.configLoader)));
  handler.register('disable_channel', withAdminGuard('disable_channel', makeDisableChannel(deps.configLoader)));
}

// ---------------------------------------------------------------------------
// Admin guard
// ---------------------------------------------------------------------------

/**
 * Asserts that the tool call originates from the main team context.
 * Admin tools must only be callable by the main assistant.
 * Throws AccessDeniedError if context is missing or teamSlug is not 'main'.
 */
function assertAdminContext(context: ToolCallContext | undefined, toolName: string): void {
  if (context === undefined || context.teamSlug !== 'main') {
    throw new AccessDeniedError(
      'tool',
      `${toolName} is restricted to the main assistant`,
    );
  }
}

/**
 * Wraps a ToolFunc with an admin guard that requires main team context.
 */
function withAdminGuard(toolName: string, fn: ToolFunc): ToolFunc {
  return async (args: Record<string, JsonValue>, context?: ToolCallContext): Promise<JsonValue> => {
    assertAdminContext(context, toolName);
    return fn(args, context);
  };
}

// ---------------------------------------------------------------------------
// Uptime formatting
// ---------------------------------------------------------------------------

/**
 * Formats a duration in milliseconds as a human-readable string.
 * Examples: "45s", "2m 15s", "1h 3m 7s"
 */
function formatUptime(ms: number): string {
  const totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;

  const parts: string[] = [];
  if (hours > 0) parts.push(`${hours}h`);
  if (minutes > 0) parts.push(`${minutes}m`);
  parts.push(`${seconds}s`);
  return parts.join(' ');
}

// ---------------------------------------------------------------------------
// Redaction helpers
// ---------------------------------------------------------------------------

/**
 * Returns a deep copy of ChannelsConfig with sensitive token fields redacted.
 */
function redactChannels(channels: MasterConfig['channels']): MasterConfig['channels'] {
  return {
    discord: {
      ...channels.discord,
      token: channels.discord.token !== undefined && channels.discord.token !== ''
        ? '[REDACTED]'
        : channels.discord.token,
    },
    whatsapp: {
      ...channels.whatsapp,
      token: channels.whatsapp.token !== undefined && channels.whatsapp.token !== ''
        ? '[REDACTED]'
        : channels.whatsapp.token,
    },
  };
}

// ---------------------------------------------------------------------------
// get_config
// ---------------------------------------------------------------------------

/**
 * Returns a config section (or the full config) from the ConfigLoader.
 * Sensitive token fields are redacted when channels or full config is returned.
 *
 * Args:
 *   section?: string — "system" | "assistant" | "channels" | "" (full config)
 */
function makeGetConfig(configLoader: ConfigLoader): ToolFunc {
  return async (args: Record<string, JsonValue>): Promise<JsonValue> => {
    const section = args['section'] !== undefined ? args['section'] : '';

    if (typeof section !== 'string') {
      throw new ValidationError('section', 'section must be a string');
    }

    const cfg = await configLoader.loadMaster();

    switch (section) {
      case 'system':
        return cfg.system as unknown as JsonValue;

      case 'assistant':
        return cfg.assistant as unknown as JsonValue;

      case 'channels': {
        const redacted = redactChannels(cfg.channels);
        return redacted as unknown as JsonValue;
      }

      case '': {
        // Return full config with channels redacted.
        const redactedFull: MasterConfig = {
          ...cfg,
          channels: redactChannels(cfg.channels),
        };
        return redactedFull as unknown as JsonValue;
      }

      default:
        throw new ValidationError('section', `unknown section: ${section}`);
    }
  };
}

// ---------------------------------------------------------------------------
// update_config
// ---------------------------------------------------------------------------

/**
 * Updates a config field identified by section + field path, then saves.
 *
 * Args:
 *   section: string — "system" | "channels"
 *   field:   string — field path within that section
 *   value:   JsonValue — new value (string or boolean depending on field)
 */
function makeUpdateConfig(configLoader: ConfigLoader): ToolFunc {
  return async (args: Record<string, JsonValue>): Promise<JsonValue> => {
    const section = args['section'];
    const field = args['field'];
    const value = args['value'];

    if (typeof section !== 'string' || section === '') {
      throw new ValidationError('section', 'section is required');
    }
    if (typeof field !== 'string' || field === '') {
      throw new ValidationError('field', 'field is required');
    }
    if (value === undefined || value === null) {
      throw new ValidationError('value', 'value is required');
    }

    const cfg = await configLoader.loadMaster();

    switch (section) {
      case 'system':
        applySystemUpdate(cfg, field, value);
        break;
      case 'channels':
        applyChannelUpdate(cfg, field, value);
        break;
      default:
        throw new ValidationError('section', `unsupported section for update: ${section}`);
    }

    await configLoader.saveMaster(cfg);

    return { status: 'updated' };
  };
}

/**
 * Applies a field update to the system config section.
 */
function applySystemUpdate(cfg: MasterConfig, field: string, value: JsonValue): void {
  switch (field) {
    case 'log_level':
      if (typeof value !== 'string') {
        throw new ValidationError('value', 'log_level must be a string');
      }
      cfg.system.log_level = value;
      break;

    case 'listen_address':
      if (typeof value !== 'string') {
        throw new ValidationError('value', 'listen_address must be a string');
      }
      cfg.system.listen_address = value;
      break;

    default:
      throw new ValidationError('field', `unsupported system field: ${field}`);
  }
}

/**
 * Applies a field update to the channels config section.
 */
function applyChannelUpdate(cfg: MasterConfig, field: string, value: JsonValue): void {
  switch (field) {
    case 'discord.enabled':
      if (typeof value !== 'boolean') {
        throw new ValidationError('value', 'discord.enabled must be a boolean');
      }
      cfg.channels.discord.enabled = value;
      break;

    case 'whatsapp.enabled':
      if (typeof value !== 'boolean') {
        throw new ValidationError('value', 'whatsapp.enabled must be a boolean');
      }
      cfg.channels.whatsapp.enabled = value;
      break;

    default:
      throw new ValidationError('field', `unsupported channel field: ${field}`);
  }
}

// ---------------------------------------------------------------------------
// get_system_status
// ---------------------------------------------------------------------------

/**
 * Returns current system status: connected teams, uptime, version.
 */
function makeGetSystemStatus(deps: AdminToolsDeps): ToolFunc {
  return async (_args: Record<string, JsonValue>): Promise<JsonValue> => {
    const teams = deps.wsHub.getConnectedTeams();
    const uptimeMs = Date.now() - deps.startTime.getTime();

    return {
      connected_teams: teams,
      uptime: formatUptime(uptimeMs),
      version: '0.1.0',
    };
  };
}

// ---------------------------------------------------------------------------
// list_channels
// ---------------------------------------------------------------------------

/**
 * Returns the list of currently connected team IDs.
 */
function makeListChannels(wsHub: WSHub): ToolFunc {
  return async (_args: Record<string, JsonValue>): Promise<JsonValue> => {
    const teams = wsHub.getConnectedTeams();
    return { connected_teams: teams };
  };
}

// ---------------------------------------------------------------------------
// enable_channel / disable_channel
// ---------------------------------------------------------------------------

/**
 * Enables a named messaging channel in the config, then saves.
 */
function makeEnableChannel(configLoader: ConfigLoader): ToolFunc {
  return async (args: Record<string, JsonValue>): Promise<JsonValue> => {
    const channel = args['channel'];
    if (typeof channel !== 'string' || channel === '') {
      throw new ValidationError('channel', 'channel name is required');
    }

    const cfg = await configLoader.loadMaster();

    switch (channel) {
      case 'discord':
        cfg.channels.discord.enabled = true;
        break;
      case 'whatsapp':
        cfg.channels.whatsapp.enabled = true;
        break;
      default:
        throw new ValidationError('channel', `unknown channel: ${channel}`);
    }

    await configLoader.saveMaster(cfg);
    return { status: 'enabled', channel };
  };
}

/**
 * Disables a named messaging channel in the config, then saves.
 */
function makeDisableChannel(configLoader: ConfigLoader): ToolFunc {
  return async (args: Record<string, JsonValue>): Promise<JsonValue> => {
    const channel = args['channel'];
    if (typeof channel !== 'string' || channel === '') {
      throw new ValidationError('channel', 'channel name is required');
    }

    const cfg = await configLoader.loadMaster();

    switch (channel) {
      case 'discord':
        cfg.channels.discord.enabled = false;
        break;
      case 'whatsapp':
        cfg.channels.whatsapp.enabled = false;
        break;
      default:
        throw new ValidationError('channel', `unknown channel: ${channel}`);
    }

    await configLoader.saveMaster(cfg);
    return { status: 'disabled', channel };
  };
}
